Allowing inferred searching using facets as keywords

bistline opened the pull request:
Non-numeric faceted searches will now attempt to "infer" results by converting filter values into exact phrases (using filter display names, e.g. "tuberculosis", or "Homo sapiens") and running a secondary search after all exact matches are found.  These studies are labeled with `inferred_match: true` in the response, and ordered after all exact matches have returned.  They will also list what terms matched using the standard `term_matches` parameter.  These results are weighted together, so higher confidence inferred matches are returned first (but still after all exact matches).  This PR also DRYs up the search logic somewhat by extracting MongoDB-based searches into a static method.

This PR satisfies SCP-2205.

devonbush commented on app/controllers/api/v1/search_controller.rb:
```diff
@@ -557,6 +574,27 @@ def self.escape_terms_for_regex(term_list:)
+      # and :inferred (converting a facet-based query to keywords)
+      # will scope the query based off of :base_studies, and include/exclude studies matching
+      # :accessions based on the :query_case (included by default, but excluded in :inferred to avoid duplicates)
+      def self.generate_mongo_query_by_case(terms:, base_studies:, accessions:, query_case:)
```
hmmm.. I'm trying to think of a better word to use than 'case', since I at first thought this was referring to whether the search was case-sensitive. 'type'? 'term_type'? 'goal'? I'm not sure any of those are better though

bistline replied:
I landed on `generate_mongo_query_by_context`, and changed the last parameter to `query_context`.

devonbush replied:
nice

devonbush commented on app/controllers/api/v1/search_controller.rb:
```diff
@@ -612,6 +650,19 @@ def self.generate_bq_query_string(facets)
+        filter_terms = []
+        facets.each do |facet|
+          search_facet = SearchFacet.find(facet[:object_id])
+          # only use non-numeric facets
```
nice catch :). That's definitely an edge case I would have missed

devonbush commented on app/controllers/api/v1/search_controller.rb:
```diff
@@ -612,6 +650,19 @@ def self.generate_bq_query_string(facets)
+          search_facet = SearchFacet.find(facet[:object_id])
+          # only use non-numeric facets
+          if !search_facet.is_numeric?
+            filter_terms += search_facet[:filters].map {|filter| filter[:name]}
```
I think this can be rewritten as 
```suggestion
            filter_terms += search_facet.filters.map(&:name)
```

bistline replied:
Actually it cannot.  The shorthand block notation does not work with hash keys (I wish it did).

devonbush commented on test/api/search_controller_test.rb:
```diff
@@ -245,4 +245,38 @@ class SearchControllerTest < ActionDispatch::IntegrationTest
+    other_study.update(description: original_description)
+
+    puts "#{File.basename(__FILE__)}: #{self.method_name} successful!"
+  end
```
This is good.  Can you also add a test for 
1. Searching for multiple facets/filters behaves as expected
2. Searching for an exact phrase + facet behaves as expected.  
I expect both of these cases to work, but given that this is home-page behavior, an extra layer of testing would be good

bistline replied:
Can do.

bistline replied:
To be clear, we already have a test for the first case.  The existing test `'should return search results using facets'` uses multiple facets at the same time.  But I can add a test for the second case.

bistline replied:
70f04a9

bistline commented:
> I don't know what our conventions/log setups are, but this feels more like a trace/debug log than INFO

I hear you, but to be perfectly honest the logic behind searches now is somewhat confusing.  The presence of keyword + facets + inferred search led to some corner cases that I didn't expect.  So those statements started out as debug, but then I decided they will be useful for addressing issues in production where we think we should be getting results but might not.

devonbush commented:
> I don't see the upside to that, really. We're trying to test the corner cases of inferred search, so I don't see the difference between extra assertions in a test that is already targeting that feature and extracting it out to its own test. I'd have to do the same setup that I did in this test.

the benefit is that when something breaks, you get a full picture of what's broken and still working, rather than only learning what the first failure was.  Concur that it's non-essential in this case

bistline commented:
> > I don't see the upside to that, really. We're trying to test the corner cases of inferred search, so I don't see the difference between extra assertions in a test that is already targeting that feature and extracting it out to its own test. I'd have to do the same setup that I did in this test.
> 
> the benefit is that when something breaks, you get a full picture of what's broken and still working, rather than only learning what the first failure was. Concur that it's non-essential in this case

Fair enough - I'll abstract it out to a separate test
